src/pcre2_util.h

```diff
@@ -52,9 +52,7 @@ POSSIBILITY OF SUCH DAMAGE.
 #endif
 
 #ifdef PCRE2_DEBUG
-#if defined(HAVE_BUILTIN_EXPECT) && defined(HAVE_BUILTIN_UNREACHABLE)
-#define PCRE2_ASSERT(x) do { if (__builtin_expect(!(x), 0)) __builtin_unreachable(); } while (0)
-#elif defined(HAVE_ASSERT_H)
+#if defined(HAVE_ASSERT_H) && !defined(NDEBUG)
```

Comment:
Would it make sense for the non-`PCRE2_DEBUG` case to do use the old (removed in this commit) version of `PCRE2_ASSERT` there?

Reply:
I don't think so, what is that you would like to accomplish by doing so?

Reply:
I'm not really bothered about it, I just wanted to note that it would let you keep the optimisation benefits in non-debug builds and also get a trap from UBSAN when asserts are disabled. If it's not desired, no problem.

Reply:
I see; `PCRE2_ASSERT()` doesn't generate any code right now in non debug builds, so it is already optimized for speed AFAIK.

Reply:
I prefer no code generated by asserts in release builds. There could be release asserts if it is needed, but I have no plans for that.

Reply:
Yeah, it just means you don't get the optimisation benefit of saying "this shouldn't be reachable".

Reply:
> it just means you don't get the optimisation benefit of saying "this shouldn't be reachable"

for that there is a different macro  `PCRE2_UNREACHABLE()`, but their use is also likely to change (see #490), `PCRE2_ASSERT()` was never meant to be used to signal that, and the fact that it accidentally did because of the implementation is what is being corrected here.

Reply:
I suspect he mentions a different thing. When `(!(a < 5))` goes to unreachable, the compiler can tell that `a < 5`, and it can use it for optimization. However, I have doubts about the usefulness of such knowledge, and the condition part of the assertion may also have side effects (forcing extra memory accesses), if the compiler cannot fully eliminate them.